File: Week 1/3.Better_Hello_World.java
/*
 * WEEK 1 - JAVA FUNDAMENTALS
 * Topic: Print Statements - println vs print
 * File: 3.Better_Hello_World.java
 *
 * CONCEPT:
 * Demonstrates the difference between print() and println() methods.
 * println() appends a newline character after the output, while print()
 * does not. Multiple println() calls produce output on separate lines.
 *
 * KEY POINTS:
 * - System.out.println() adds a newline (\n) after the output
 * - System.out.print() does NOT add a newline; next output continues on same line
 * - Multiple println() calls will each print on a new line
 * - Both methods accept String arguments enclosed in double quotes
 *
 * SYNTAX:
 * System.out.println("text"); // prints with newline
 * System.out.print("text");   // prints without newline
 */

package fundamentals;
public class HelloWorld {
    public static void main(String args[]) {
// println prints a new line at the end

     System.out.println("Hello World");
     System.out.println("Hello World");
     System.out.println("Hello World");
    }
}


/* In order to print things to console we have to write - 
System.out.println("HelloWorld").
Again for now we should leave System.out.print mean, and should write it as it is The built-in method print() is used to display the string which is passed to it. This
output string is not followed by a newline, i.e., the next output will start on thesame line. 

The built-in method println() is similar to print(), except that println()  outputs a newline after each call. */ 

File: Week 1/rust/3.better_hello_world.rs
/*
 * WEEK 1 - RUST FUNDAMENTALS
 * Topic: print! / println! / eprintln! / format!
 * File: 3.better_hello_world.rs
 *
 * CONCEPT:
 * Rust offers a family of formatting macros instead of separate `print` and
 * `println` methods:
 *  - print!     -> stdout, no newline
 *  - println!   -> stdout, with newline
 *  - eprint!    -> stderr, no newline
 *  - eprintln!  -> stderr, with newline
 *  - format!    -> returns a String instead of printing
 *
 * KEY POINTS:
 *  - Macros use `!` syntax: println!(...)
 *  - Format string uses `{}` for positional args (Display trait) or `{:?}` for Debug
 *  - Width, precision, alignment: `{:>10}`, `{:.3}`, `{:0>5}`, etc.
 *  - Named arguments: `println!("{name} {age}", name="Bob", age=30)`
 *  - From Rust 1.58+ you can also CAPTURE local variables directly: `println!("{x}")`.
 *
 * SYNTAX:
 *  println!("hello");
 *  print!("no newline");
 *  let s = format!("x = {}", 42);   // returns a String
 *
 * DRY RUN:
 *  Output 1: three "Hello World" on three lines.
 *  Output 2: three on the same line (using print!).
 *  Output 3: a formatted string built with format!.
 */

fn main() {
    // println!-style: three lines
    println!("Hello World");
    println!("Hello World");
    println!("Hello World");

    println!();   // empty newline

    // print!-style: same line
    print!("Hello World");
    print!("Hello World");
    print!("Hello World");
    println!();   // close the line

    println!();

    // format! returns a String (no I/O)
    let greeting: String = format!("Hello, {}!", "Rust");
    println!("{}", greeting);

    // Named captures (Rust 1.58+)
    let x = 7;
    let y = 11;
    println!("x = {x}, y = {y}, x+y = {}", x + y);
}

/*
 * NOTES:
 *  - Java has println / print methods on System.out / System.err.
 *  - Python has one `print()` function with an `end=` keyword arg.
 *  - Rust splits the choice into separate macros (`print!`, `println!`, ...).
 *  - `format!` is the analogue of `String.format(...)` (Java) or f-strings (Python).
 *  - For binary printing use `{:b}`, for hex `{:x}`, for debug-pretty `{:#?}`.
 */
